### PR TITLE
Add flag to enable logging on rejected gossip message

### LINKED ISSUE #11431: Add option to log full ssz data on rejected gossip messages
Learned from @ajsutton that this is what Teku does and it's very handy, for example we remove invalid blocks from DB, we may keep the data in the logs.

### PR DESCRIPTION
**Pull request by saolyn**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

This PR adds a new feature flag to enable logging on rejected gossip messages.

**Which issues(s) does this PR fix?**

Fixes #11431

**Other notes for review**

If we add this flag to the docs we need to make sure we add a note about the potential risks of running a node with this flag enabled as pointed out on the issue.
Essentially that this can easily become a potential DoS vector.

### CODE REVIEW

File: beacon-chain/sync/subscriber.go [hunk 1]
```diff
@@ -765,3 +766,10 @@ func multiAddr(pid peer.ID, stat *peers.Status) string {
 	}
 	return addrs.String()
 }
+
+func rejectGossipMessage(msg *pubsub.Message) pubsub.ValidationResult {
+	if features.Get().EnableFullSSZDataLogging {
+		log.WithField("rawMessage", msg).Debug("Rejected gossip message with data")
+	}
+	return pubsub.ValidationReject
+}
```
Comment (saolyn) [inline beacon-chain/sync/subscriber.go:775]: I am not entirely sure if this is the best place for this function.

File: beacon-chain/sync/validate_aggregate_proof_test.go [hunk 2]
```diff
@@ -606,7 +606,7 @@ func TestValidateAggregateAndProof_BadBlock(t *testing.T) {
 	}
 	res, err := r.validateAggregateAndProof(context.Background(), "", msg)
 	assert.NotNil(t, err)
-	assert.Equal(t, pubsub.ValidationReject, res, "Validated status is true")
+	assert.Equal(t, rejectGossipMessage(msg), res, "Validated status is true")
 }
```
Comment (prestonvanloon) [inline beacon-chain/sync/validate_aggregate_proof_test.go:610]: No need to change the tests. Lets revert all of these test changes so we can ensure the behavior stays the same

File: beacon-chain/sync/subscriber.go (hunk 1 above)
Comment (prestonvanloon) [inline beacon-chain/sync/subscriber.go:775]: Could you add a function level comment? 

What is the intended behavior of this function?

File: beacon-chain/sync/subscriber.go [hunk 3]
```diff
@@ -3,6 +3,7 @@ package sync
 import (
 	"context"
 	"fmt"
+	"github.com/prysmaticlabs/prysm/v3/config/features"
 	"reflect"
 	"runtime/debug"
 	"strings"
```
Comment (prestonvanloon) [inline beacon-chain/sync/subscriber.go:9]: Please put this import with the other non standard library imports, below line 12

**Comment by prestonvanloon** [comment]
If you are trying to log the message as a ssz hex string, I don't think printing the `*pubsub.Message` would work. Could you add a test to ensure the behavior is as expected?

File: beacon-chain/sync/subscriber.go [hunk 4]
```diff
@@ -765,3 +766,10 @@ func multiAddr(pid peer.ID, stat *peers.Status) string {
 	}
 	return addrs.String()
 }
+
+func rejectGossipMessage(msg *pubsub.Message) pubsub.ValidationResult {
+	if features.Get().EnableFullSSZDataLogging {
+		log.WithField("rawMessage", msg).Debug("Rejected gossip message with data")
```
Comment (prestonvanloon) [inline beacon-chain/sync/subscriber.go:772]: If you want to match Teku's log, they also log the reason or error.

https://github.com/ConsenSys/teku/blob/cdf161742a7abb091ab04fb6c23fee622b9d1df7/infrastructure/logging/src/main/java/tech/pegasys/teku/infrastructure/logging/P2PLogger.java#L50

**Comment by potuz** [comment]
> If you are trying to log the message as a ssz hex string, I don't think printing the `*pubsub.Message` would work. Could you add a test to ensure the behavior is as expected?

Yeah I think you want to convert this to a hex string that can be directly fed to a ssz marshaller. Also notice that it's mandatory to have the expected object at least, otherwise we won't know what schema to use to unmarshal

**Comment by prestonvanloon** [comment]
> > If you are trying to log the message as a ssz hex string, I don't think printing the `*pubsub.Message` would work. Could you add a test to ensure the behavior is as expected?
> 
> Yeah I think you want to convert this to a hex string that can be directly fed to a ssz marshaller. Also notice that it's mandatory to have the expected object at least, otherwise we won't know what schema to use to unmarshal

I _think_ `msg.Data` is the ssz payload? It might be snappy compressed, i don't remember, but it's the raw message payload

**Review by nisdas — CHANGES_REQUESTED** [review]
This isnt the correct place to check for it, it needs to be in our gossip handler wrapper.

**Comment by potuz** [comment]
> I _think_ `msg.Data` is the ssz payload? It might be snappy compressed, i don't remember, but it's the raw message payload
> 
> https://pkg.go.dev/github.com/libp2p/go-libp2p-pubsub@v0.8.1/pb#Message

Yes but if you print a byte slice with the logger it won't print it as a hex string, and also to decode the ssz you need to know what object the message was.

File: beacon-chain/sync/validate_aggregate_proof_test.go (hunk 2 above)
Comment (saolyn) [reply beacon-chain/sync/validate_aggregate_proof_test.go:610]: good catch, I completely missed this one.

File: beacon-chain/sync/subscriber.go (hunk 3 above)
Comment (saolyn) [reply beacon-chain/sync/subscriber.go]: fixed

File: beacon-chain/sync/subscriber.go [hunk 5]
```diff
@@ -257,13 +259,19 @@ func (s *Service) wrapAndReportValidation(topic string, v wrappedVal) (string, p
...
 				"gossip score": s.cfg.p2p.Peers().Scorers().GossipScorer().Score(pid),
-			}).Debugf("Gossip message was rejected")
+			}
+			if features.Get().EnableFullSSZDataLogging {
+				obj := p2p.GossipTopicMappings(topic, slots.ToEpoch(s.cfg.chain.CurrentSlot()))
+				fields["object"] = fmt.Sprintf("%T", obj)
+				fields["message"] = hexutil.Encode(msg.Data)
+			}
+			log.WithError(err).WithFields(fields).Debugf("Gossip message was rejected")
 			messageFailedValidationCounter.WithLabelValues(topic).Inc()
```
[the remark this answers is not in the document]
Comment (nisdas) [reply beacon-chain/sync/subscriber.go:275]: Yes this is better, however you do not need to print out the object. We can infer it directly from the topic, so you can remove that one.

[the remark this answers is not in the document]
Comment (saolyn) [reply beacon-chain/sync/subscriber.go]: removed it!

**Review by nisdas — APPROVED** [review]
LGTM

**Review by rauljordan — APPROVED** [review]

**Pull request merged**